File: cs/competitive_programming/leetcode_fulltimes/2023/june/141_linked_list_cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// Procedure 1:

#include <set>

class Solution {
public:
    bool hasCycle(ListNode *head) {

        ListNode* ptr = head;
        std::set<ListNode*> ptr_set;

        for(; ptr!=nullptr; ptr=ptr->next) {

            auto it = ptr_set.find(ptr);

            // ptr present in the set
            if(it!=ptr_set.end()){
                return true;
            } else {
                ptr_set.insert(ptr);
            }
        }

        return false;
    }
};

// Procedure 2:

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {

        if(head==nullptr || head->next==nullptr) {
            return false;
        }

        ListNode* s = head; //slow ptr
        ListNode* f = head->next; // fast ptr

        for(; (f!=nullptr && f->next!=nullptr && f!=s) ; s=s->next, f=f->next->next) {
            ;
        }

        if(s==f)
            return true;
        else
            return false;

    }
};

File: cs/competitive_programming/leetcode_interns/LinkedList/EASY/141_linked_list_cycle.py
#!/usr/bin/python3

# Link - https://leetcode.com/problems/linked-list-cycle/

# Definition for singly-linked list.

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        
        if head != None and head.next != None:
            sp,fp = head,head.next
        else:
            return False
        
        while(sp != None and sp.next != None and fp.next != None and fp.next.next != None):
            
            sp = sp.next
            fp = fp.next.next
            
            if sp == fp:
                return True
        return False

if __name__ == "__main__":
    
    obj = Solution()
    
    head = ListNode(3)
    head1 = ListNode(2)
    head2 = ListNode(0)
    head3 = ListNode(-4)
    
    head.next = head1
    head1.next = head2
    head2.next = head3
    # head3.next = head1

    # head = ListNode(1)
    # head1 = ListNode(2)
    # head.next = head1


    out = obj.hasCycle(head)
    print("out is {}".format(out))
        